File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        """
        :type asteroids: List[int]
        :rtype: List[int]
        """
        stack = []

        for a in asteroids:
            if a > 0:
                stack.append(a)
            else:
                while stack and stack[-1] > 0 and stack[-1] < -a:
                    stack.pop()
                if not stack or stack[-1] < 0:
                    stack.append(a)
                elif stack and stack[-1] == -a:
                    stack.pop()

        return stack

File: 0735-asteroid-collision/0735-asteroid-collision.go
func asteroidCollision(asteroids []int) []int {
    stack := []int{}
    for i := 0; i < len(asteroids); i++ {
        if asteroids[i] > 0 {
            stack = append(stack, asteroids[i])
        } else {
            if len(stack) == 0 || stack[len(stack)-1] < 0 {
                stack = append(stack, asteroids[i])
                continue
            }

            if stack[len(stack)-1] == -asteroids[i] {
                stack = stack[:len(stack)-1]
            } else if stack[len(stack)-1] < -asteroids[i] {
                stack = stack[:len(stack)-1]
                i -= 1
            }
        }
    }

    return stack
}